plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -1454,115 +1559,40 @@ int amdgpu_plugin_restore_file(int id)
-				/* direct memcpy is possible on large bars */
-				memcpy(addr, (void *)e->bo_info_test[i]->bo_rawdata.data,
-				       (e->bo_info_test[i])->bo_size);
-				munmap(addr, e->bo_info_test[i]->bo_size);
-			} else {
-				int mem_fd;
-				/* Use indirect host data path via /proc/pid/mem
-				 * on small pci bar GPUs or for Buffer Objects
-				 * that don't have HostAccess permissions.
-				 */
-				plugin_log_msg("amdgpu_plugin: using PROCPIDMEM to restore BO contents\n");
-				addr = mmap(NULL,
-					    (e->bo_info_test[i])->bo_size,
-					    PROT_NONE,
-					    MAP_SHARED,
-					    drm_render_fd,
-					    restored_bo_offsets_array[i]);
-				if (addr == MAP_FAILED) {
-					pr_perror("amdgpu_plugin: mmap failed\n");
-					fd = -EBADFD;
-					goto clean;
-				}
-
-				if (asprintf (&fname, PROCPIDMEM, e->pid) < 0) {
-					pr_perror("failed in asprintf, %s\n", fname);
-					munmap(addr, e->bo_info_test[i]->bo_size);
-					fd = -EBADFD;
-					goto clean;
-				}
-
-				mem_fd = open (fname, O_RDWR);
-				if (mem_fd < 0) {
-					pr_perror("Can't open %s for pid %d\n", fname, e->pid);
-					free (fname);
-					munmap(addr, e->bo_info_test[i]->bo_size);
-					fd = -EBADFD;
-					goto clean;
-				}
-
-				plugin_log_msg("Opened %s file for pid = %d\n", fname, e->pid);
-				free (fname);
+	for (int i = 0; i < e->num_of_gpus; i++) {
+		int ret_thread = 0;
 
-				if (lseek (mem_fd, (off_t) addr, SEEK_SET) == -1) {
-					pr_perror("Can't lseek for bo_offset for pid = %d\n", e->pid);
-					munmap(addr, e->bo_info_test[i]->bo_size);
-					fd = -EBADFD;
-					goto clean;
-				}
+		thread_datas[i].gpu_id = devinfo_bucket_ptr[i].actual_gpu_id;
+		thread_datas[i].bo_buckets = bo_bucket_ptr;
+		thread_datas[i].bo_info_test = e->bo_info_test;
+		thread_datas[i].pid = e->pid;
+		thread_datas[i].num_of_bos = e->num_of_bos;
+		thread_datas[i].restored_bo_offsets = restored_bo_offsets_array;
+		thread_datas[i].drm_fd = devinfo_bucket_ptr[i].drm_fd;
 
-				plugin_log_msg("Attempt writting now\n");
-				if (write(mem_fd, e->bo_info_test[i]->bo_rawdata.data,
-					  (e->bo_info_test[i])->bo_size) !=
-				    (e->bo_info_test[i])->bo_size) {
-					pr_perror("Can't write buffer\n");
-					munmap(addr, e->bo_info_test[i]->bo_size);
-					fd = -EBADFD;
-					goto clean;
-				}
-				munmap(addr, e->bo_info_test[i]->bo_size);
-				close(mem_fd);
-			}
-		} else {
-			pr_info("Not a VRAM BO\n");
-			continue;
+		ret_thread = pthread_create(&thread_datas[i].thread, NULL, restore_bo_contents, (void*) &thread_datas[i]);
+		if (ret_thread) {
+			pr_err("Failed to create thread[%i]\n", i);
+			fd = -EBADFD;
+			goto clean;
 		}
-	} /* mmap done for VRAM BO */
+	}
 
 	for (int i = 0; i < e->num_of_gpus; i++) {
+		pthread_join(thread_datas[i].thread, NULL);
+		pr_info("Thread[0x%x] finished ret:%d\n", thread_datas[i].gpu_id, thread_datas[i].ret);
+
 		if (devinfo_bucket_ptr[i].drm_fd >= 0)
 			close(devinfo_bucket_ptr[i].drm_fd);
+
+		if (thread_datas[i].ret) {
+			fd = -EBADFD;
```

Comment:
Assign an appropriate error code in the thread function. Then you don't have to make one up here.